### PR TITLE
[NoQA] [Audit][Implementation] - Increase max cache keys count and use waitForCollectionCallback

### PR DESCRIPTION
**Pull request by hurali97**
### Details

This PR is part of the implementation phase of App Startup Audit by Callstack.

Originally the proposal is [here](https://github.com/Expensify/App/issues/35234#issuecomment-1926913338) but after thinking it thoroughly, we decided to follow a better solution.

We have lots of connections to `Reports` data in Onyx in our utils files and some of them uses `waitForCollectionCallback: true` and the rest of them doesn't. Which adds up to the app startup time. The reason is that when we don't set `waitForCollectionCallback` to `true` in `Onyx.js` the following code is executed:

```ts
if (mapping.waitForCollectionCallback) {
    getCollectionDataAndSendAsObject(matchingKeys, mapping);
    return;
}

 // We did not opt into using waitForCollectionCallback mode so the callback is called for every matching key.
for (let i = 0; i < matchingKeys.length; i++) {
    get(matchingKeys[i]).then((val) => sendDataToConnection(mapping, val, matchingKeys[i], true));
}
```

This loop will be executed for all the keys and will individually inform the subscriber of updates. The `get` function first checks whether the value for a key is present in cache and if it doesn't it then gets it from the storage.

We have 6 connections to `Reports` without the `waitForCollectionCallback` so this loop will be executed 6 times. Bearing in mind that `Reports` can be a huge collection, for eg on a heavy account it can get to 15k reports.

The solution is to refactor our current `Onyx.connect` calls to `Reports` to include `waitForCollectionCallback` and adjust the code accordingly. There's one place where we won't be changing it and that's in `ReportUtils.ts` where we have to call `handleReportChanged` on each report.

One more thing to notice is that currently our cache limit is 10k which is easily hit by a heavy account of 15k reports, so this cache limit isn't helping the app in case of a heavy account. We can increase it to 20k so that it can hold more reports.

Below are Hermes traces that have been gathered on:

- a real android device
- with release variant of Expensify App
- on a heavy account

<details>
<summary>Baseline</summary>

baseline-32ms

</details>

<details>
<summary>With waitForCollectionCallback</summary>`

improvement-25ms

</details>

<details>
<summary>With increased cache and waitForCollectionCallback</summary>

cache-increase-18ms

</details>

To summarise, When launching the app on Android:

- We get 32 seconds on baseline, which is what we currently have 🔴 
- 25 seconds with improvements by adding `waitForCollectionCallback` 🟢 
- 18 seconds with increased cache to 20k and improvements by adding `waitForCollectionCallback` 🟢 🟢 

### Fixed Issues

$ https://github.com/Expensify/App/issues/38586
PROPOSAL:

### Tests

- [X] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47336142/e8f6fa23-527f-49d5-9fcc-68246a2ff4f9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/47336142/08e1cae7-c962-48b4-a597-4daba2154906

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/47336142/fd0e2a1b-d852-4a81-95da-22fbf71d23c6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/47336142/454ab4ca-4b65-4e89-a575-b7f3570ce7c4

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47336142/a02fd2e8-cd7e-4aac-9b3d-1f4611c208a8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47336142/2e490e1e-fd41-443f-9cfe-dd3640766181

</details>

### CODE REVIEW

**Review by kacper-mikolajczak — APPROVED** [review]
LGTM 👍 

Question @hurali97, what's stopping us from having single connection to `ONYXKEYS.COLLECTION.REPORT`, which would result in single instance of `reports` object that then later could be shared across the app?

**Comment by hurali97** [comment]
> what's stopping us from having single connection to `ONYXKEYS.COLLECTION.REPORT`, which would result in single instance of `reports` object that then later could be shared across the app?

This is more complicated then it looks like. A straight forward solution is to implement something like [this](https://github.com/Expensify/App/issues/35234#issuecomment-1926913338). This will keep only two connections to Onyx, one for `waitForCollectionCallback` and second without it. It uses a singleton based approach with event driven updates. 

However, this will get more complicated over the time as we'll have to do the same for each of the collection, for eg, `ReportAction`.

I also evaluated whether we can update some parts of Onyx to have only one connection per key but that's easier said than done. The reason is that we have callbacks with each connection and even if we manage to keep only one connection per key, we will have to store the callbacks like below:

```js
{
key: 'report',
callbacks: [
  {
   callback: () => {},
   waitForConnectionCallback: false,
   ...otherProps,
  },
  {
    callback: () => {},
    waitForConnectionCallback: true,
    ...otherProps,
  }
 ]
}
```

Even with above structure, we will have to loop over the callbacks to fire them whenever `Report` changes.

Now, given all of the above I think it's better to avoid the individual loop within the Onyx for the larger keys. Which means reduce the chances of iterating over the keys and getting them from cache, then from storage by using `waitForCollectionCallback: true`.

**Comment by kacper-mikolajczak** [comment]
What I had in mind is starting small. The _**passive usages**_ of `Report` collection, the ones where we `connect` and callback only serves us to dump fresh data into local variable. An example of it:
https://github.com/Expensify/App/blob/bbf4bee334adff04a585ab81f1b30364e26fc84c/src/libs/OptionsListUtils.ts#L251-L259
 
 Could be replaced with single instance which we could import in many places. This would reduce code complexity as all passive connection would be explicitly handled in one place, e.g. `OnyxStore.ts` file.
 
 For _**interactive usages**_ (where live callback is needed), then we could implement it with Observer pattern like so (presented in simple form for brevity):
 
 ```ts
 const store = {};
 const listeners = new Set();

Onyx.connect({
    key: 'some_key',
    callback: (data) => {
        store = data;
        listeners.forEach(l => l(store));
    },
});

const addListener = (l) => { 
    listeners.add(l);
    return () => listeners.remove(l)
}

...

export { addListener, removeListener, data }
```

This is more complex topic but the main take-away is that, as you pointed out in your analysis, we have many redundant connections scattered across the app which builds up both mental and perf overheads.

**Comment by hurali97** [comment]
@kacper-mikolajczak I see. I implemented something similar when working in analysis phase [here](https://github.com/callstack-internal/Expensify-App/commit/b58de104543ed43e0097607e289ee2bea6e6e2d1). What you’ve proposed is more or less the same thing.

The only issue is the complexity that this solution brings. Just see that we will have to do this for each key, if we want to use it. Also, in my profiling with the changes in the PR, I realise that this single connection approach doesn’t bring much of an improvement. So I guess with these improvements, this doesn’t make sense to introduce more complexity just for the sake of single connection because it’s not hurting the performance, courtesy of the changes added in the PR now :thinking_face:

Previously, what you’ve mentioned was my go to approach but I guess we have something better now, so I will keep it until in the future we encounter it again then we can continue the discussion.

**Comment by kacper-mikolajczak** [comment]
Great, thanks for the explanation! We can definitely continue the discussion :)

File: src/libs/OptionsListUtils.ts
Comment (roryabraham) [inline src/libs/OptionsListUtils.ts:255]: same here, this should be `allReports`

File: src/libs/OptionsListUtils.ts
Comment (roryabraham) [reply src/libs/OptionsListUtils.ts:251]: or filter the reports in the `Onyx.connect` callback?

```ts
let policyExpenseReports: OnyxCollection<Report> = {};
Onyx.connect({
    key: ONYXKEYS.COLLECTION.REPORT,
    waitForCollectionCallback: true,
    callback: (allReports) => {
        policyExpenseReports = Object.entries(allReports)
            .filter(([reportID, report]) => ReportUtils.isPolicyExpenseChat(report))
            .reduce((memo, [reportID, report]) => {
                memo[reportID] = report;
                return memo;
            }, {});
    }
});
```

File: src/libs/actions/Report.ts
Comment (roryabraham) [inline src/libs/actions/Report.ts:153]: can we avoid suppressing this lint rule?

File: src/libs/OptionsListUtils.ts
Comment (roryabraham) [reply src/libs/OptionsListUtils.ts:251]: note that in this snippet ^ I'm avoiding using a spread in a reduce, which is very inefficient, especially for a potentially large object

Comment (hurali97) [reply src/libs/OptionsListUtils.ts:251]: @roryabraham @shubham1206agra 

On a second thought, I think we should remove the `Onyx.connect` here. The reason is that it's only used to populate `policyExpenseReports` which is only used in `getPolicyExpenseReportOption`. I looked for the usages of this function throughout the codebase and it turns out that the `report` we pass to this function is already verified as a policy expense report, so it is not ideal to get the report again from the collection we store from `Onyx.connect` and then access the report from collection by using the `reportID` passed by the report object to the `getPolicyExpenseReportOption`.

Screenshot 2024-03-20 at 5 03 57 PM


<br/>
<hr/>
Below is how the changes will look like:
<br/>

```diff
diff --git a/src/libs/OptionsListUtils.ts b/src/libs/OptionsListUtils.ts
index 73620a2b6f..ab132c21c2 100644
--- a/src/libs/OptionsListUtils.ts
+++ b/src/libs/OptionsListUtils.ts
@@ -248,15 +248,6 @@ Onyx.connect({
     },
 });
 
-let policyExpenseReports: OnyxCollection<Report> = {};
-Onyx.connect({
-    key: ONYXKEYS.COLLECTION.REPORT,
-    waitForCollectionCallback: true,
-    callback: (report) => {
-        policyExpenseReports = report;
-    },
-});
-
 let allTransactions: OnyxCollection<Transaction> = {};
 Onyx.connect({
     key: ONYXKEYS.COLLECTION.TRANSACTION,
@@ -735,12 +726,10 @@ function createOption(
  * Get the option for a policy expense report.
  */
 function getPolicyExpenseReportOption(report: Report): ReportUtils.OptionData {
-    const expenseReport = policyExpenseReports?.[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`];
-
     const option = createOption(
-        expenseReport?.visibleChatMemberAccountIDs ?? [],
+        report?.visibleChatMemberAccountIDs ?? [],
         allPersonalDetails ?? {},
-        expenseReport ?? null,
+        report,
         {},
         {
             showChatPreviewLine: false,
@@ -749,7 +738,7 @@ function getPolicyExpenseReportOption(report: Report): ReportUtils.OptionData {
     );
 
     // Update text & alternateText because createOption returns workspace name only if report is owned by the user
-    option.text = ReportUtils.getPolicyName(expenseReport);
+    option.text = ReportUtils.getPolicyName(report);
     option.alternateText = Localize.translateLocal('workspace.common.workspace');
     option.selected = report.selected;
     option.isSelected = report.selected;

```

File: src/libs/actions/Report.ts
Comment (hurali97) [reply src/libs/actions/Report.ts:153]: @roryabraham  This is rather complex. We have `handleReportChanged` declared way down in the file and it depends on `reconnect` which is declared above `handleReportChanged`. So to avoid this lint rule, we will have to bring these two functions just above the `Onyx.connect` call. 

Which will look bad in terms of readability because as of now we have all `Onyx.connect` calls in one place and below that we have functions.

**Comment by hurali97** [comment]
@shubham1206agra any updates on the review? 🙂

**Comment by hurali97** [comment]
> @hurali97 I thought you were going to implement https://github.com/Expensify/App/pull/38464/files#r1531973454. Let me ping @roryabraham here to confirm if we can proceed with this solution.

@shubham1206agra Yeah I am willing to implement this just waiting for a green light from the reviewers 👍

File: src/libs/OptionsListUtils.ts
Comment (hurali97) [reply src/libs/OptionsListUtils.ts]: @shubham1206agra Oh yes 👍 Must have been sneaked in while resolving conflicts.

**Review by grgia — DISMISSED** [review]

File: src/libs/OptionsListUtils.ts
Comment (roryabraham) [reply src/libs/OptionsListUtils.ts]: sounds good. If we can add some inline local validation of the report object with `ReportUtils.isPolicyExpenseChat`, that could be a good safety measure without any significant performance cost

File: src/libs/actions/Report.ts
Comment (roryabraham) [reply src/libs/actions/Report.ts:139]: ok, no worries then

**Comment by hurali97** [comment]
> @hurali97 Shouldn't the same thing applies to `getReportOption`?

@shubham1206agra I think no, because the `getReportOption` is using `participant` to access the `report` from `allReports`. If `getReportOption` was passing the `report` as parameter then we could have done the same here as well.

**Comment by hurali97** [comment]
> @hurali97 You may have been misguided by the type. You may want to recheck the usage of `getPolicyExpenseReportOption`.

@shubham1206agra Yeah I was because I checked before merging the main. Anyways, things are resolved now 👍

**Comment by hurali97** [comment]
@shubham1206agra  Oh.... so you mean I should rename `report` to `participant` in `getPolicyExpenseReportOption` ?

**Comment by hurali97** [comment]
> Yes, and update the function accordingly.

I am understanding that we can rename the parameter `report` to `participant` in `getPolicyExpenseReportOption ` but what are you referring to update besides it?

Are you referring to add `ReportUtils.getReport(participant.reportID)` in `getPolicyExpenseReportOption` to get the report using this helper method? Something like below:

```ts
function getPolicyExpenseReportOption(participant: Participant): ReportUtils.OptionData {
    const expenseReport = ReportUtils.isPolicyExpenseChat(participant) ? ReportUtils.getReport(participant.reportID) : null;

....
```

**Review by shubham1206agra — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
Thank you!

**Comment by mountiny** [comment]
Going to go ahead since C+ reviewed and Georgia approved before too

**Comment by mountiny** [comment]
Melvin the gaslighter the tests were passing

File: src/libs/TaskUtils.ts
Comment (getusha) [inline src/libs/TaskUtils.ts:41]: This should've been updated to `const taskReport = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${taskReportID}`] ?? {};` after updating how we key the reports. it caused a regression https://github.com/Expensify/App/issues/40121 More information: https://github.com/Expensify/App/issues/40121#issuecomment-2050945636

**Pull request merged**
